src/transport_ofi.c

```diff
@@ -1329,6 +1333,98 @@ int allocate_fabric_resources(struct fabric_info *info)
     return ret;
 }
 
+#ifdef USE_HWLOC
+struct fi_info *assign_nic_with_hwloc(struct fi_info *fabric, struct fi_info **provs, size_t num_nics) {
+    int ret = 0;
+    hwloc_bitmap_t bindset = hwloc_bitmap_alloc();
+
+    ret = hwloc_get_proc_last_cpu_location(shmem_topology, getpid(), bindset, HWLOC_CPUBIND_PROCESS);
+    if (ret < 0) {
+        RAISE_ERROR_MSG("hwloc_get_proc_last_cpu_location failed (%s)\n", strerror(errno));
+    }
+
+    // Identify which provider entries correspond to NICs with an affinity to the calling process
+    struct fi_info *close_provs = NULL;
+    struct fi_info *last_added = NULL;
+    size_t num_close_nics = 0;
+    for (size_t i = 0; i < num_nics; i++) {
+        struct fi_info *cur_prov = provs[i];
+        if (cur_prov->nic->bus_attr->bus_type != FI_BUS_PCI) continue;
+
+        struct fi_pci_attr pci = cur_prov->nic->bus_attr->attr.pci;
+        hwloc_obj_t io_device = hwloc_get_pcidev_by_busid(shmem_topology, pci.domain_id, pci.bus_id, pci.device_id, pci.function_id);
+        if (!io_device) RAISE_ERROR_MSG("hwloc_get_pcidev_by_busid failed\n");
+        hwloc_obj_t first_non_io = hwloc_get_non_io_ancestor_obj(shmem_topology, io_device);
+        if (!first_non_io) RAISE_ERROR_MSG("hwloc_get_non_io_ancestor_obj failed\n");
+
+        if (hwloc_bitmap_isincluded(bindset, first_non_io->cpuset) ||
+            hwloc_bitmap_isincluded(first_non_io->cpuset, bindset)) {
+            struct fi_info *dup = fi_dupinfo(cur_prov);
+            if (!close_provs) close_provs = dup;
+            if (last_added) last_added->next = dup;
+            last_added = dup;
+            num_close_nics++;
+        }
+    }
+    DEBUG_MSG("Num. NICs w/ affinity to process: %zu\n", num_close_nics);
+
+    if (!close_provs) {
+        RAISE_WARN_MSG("Could not detect any NICs with affinity to the process\n");
+
+        /* If no 'close' NICs, select from list of all NICs using round-robin assignment */
+        //return provs[shmem_team_my_pe(SHMEMX_TEAM_NODE) % num_nics];
```

Comment:
I would think the node PE is required here... can it be uncommented?

If so, best to use `shmem_internal_team_node.my_pe` to preserve layering.

Reply:
I was looking into this more recently and I don't think **SHMEMX_TEAM_NODE** can be used - at least not here. The **SHMEMX_TEAM_NODE** team is initialized in **shmem_internal_team_init()**, which is not called until after **shmem_transport_init()** (which calls **query_for_fabric())** in **shmem_internal_heap_postinit()**. This explains why I was seeing that each process was indicating that it was PE 0 in SHMEMX_TEAM_NODE.

For what it's worth, a recent run of the **shmem_bw_put_perf** benchmark seemed to indicate that the round-robin indexing using SHMEM_TEAM_WORLD still worked effectively such that there was no duplicate assignment of PEs to NICs (tested from 1-8 ppn)

I'll go ahead and remove the commented out lines though.

Reply:
Oh yeah, of course...  I completely missed that, thanks.

Do you think it's possible to use this (`my_world_pe % num_nics`) form of round robin for initialization of collectives and teams, then to reassign PEs to NICs based on team-node after that?

There may be some other solution too...  With the code as written, NIC bindings will vary with the total number of PEs... so it could be difficult to understand multi-NIC performance, and seems like it will break if there are differing numbers of NICs per node.